src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs

```diff
@@ -0,0 +1,312 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Debug = System.Diagnostics.Debug;
+using System.Collections.Generic;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.TypeSystem
+{
+    public static partial class CastingHelper
+    {
+        static partial void IsEquivalentTo(this TypeDesc thisType, TypeDesc otherType, StackOverflowProtect visited, ref bool isEquivalentTo)
+        {
+            isEquivalentTo = IsEquivalentToHelper(thisType, otherType, visited);
+            return;
+        }
+
+        private static bool IsEquivalentToHelper(TypeDesc thisType, TypeDesc otherType, StackOverflowProtect visited)
+        {
+            if (thisType == otherType)
+                return true;
+
+            if (!thisType.IsTypeDefEquivalent || !otherType.IsTypeDefEquivalent)
+                return false;
+
+            if (thisType.Category != otherType.Category)
+                return false;
+
+            switch (thisType.Category)
+            {
+                case TypeFlags.SignatureTypeVariable:
+                case TypeFlags.SignatureMethodVariable:
+                case TypeFlags.GenericParameter:
+                    return false;
+
+                case TypeFlags.Array:
+                case TypeFlags.SzArray:
+                case TypeFlags.ByRef:
+                case TypeFlags.Pointer:
+                    return ((ParameterizedType)thisType).ParameterType.IsEquivalentTo(((ParameterizedType)otherType).ParameterType, visited);
+
+                case TypeFlags.FunctionPointer:
+                    return false;
+
+                default:
+                    Debug.Assert(thisType.IsDefType);
+                    // Check if type is generic
+                    return ((DefType)thisType).IsEquivalentToDefType((DefType)otherType, visited);
+            }
+        }
+
+        private static bool IsEquivalentToDefType(this DefType thisType, DefType otherType, StackOverflowProtect visited)
+        {
+            if (thisType.HasInstantiation)
+            {
+                // Limit equivalence on generics only to interfaces
+                if (!thisType.IsInterface || !otherType.IsInterface)
+                {
+                    return false;
+                }
+
+                if (thisType.Instantiation.Length != otherType.Instantiation.Length)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < thisType.Instantiation.Length; i++)
+                {
+                    if (!thisType.Instantiation[i].IsEquivalentTo(otherType.Instantiation[i], visited))
+                    {
+                        return false;
+                    }
+                }
+
+                // Generic equivalence only allows the instantiation to be non-equal
+                return thisType.HasSameTypeDefinition(otherType);
```

Comment:
Perf nit - wouldn't it be faster to first check whether the two types have the same type definition and only then traverse the instantiations?

Reply:
Good point. Changed